ipyparallel/cluster/shellcmd.py

```diff
@@ -0,0 +1,860 @@
+* running: check if a process with a given pid is running
+* kill:    kill a process with a given pid
+* mkdir:   creates directory recursively (no error if directory already exists)
+* rmdir:   remove directory recursively
+* exists:  checks if a file/directory exists
+* remove:  removes a file
+"""
+import base64
+import enum
+import inspect
+import json
+import os
+import re
+import shlex
+import sys
+import time
+from datetime import datetime
+from random import randint
+from subprocess import CalledProcessError, Popen, TimeoutExpired, check_output
+from tempfile import NamedTemporaryFile
+
+
+class Platform(enum.Enum):
+    Unknown = (0,)
+    Linux = (1,)
+    Windows = (2,)
+    MacOS = 3
+
+    @staticmethod
+    def get():
+        import sys
+
+        tmp = sys.platform.lower()
+        if tmp == "win32":
+            return Platform.Windows
+        elif tmp == "linux":
+            return Platform.Linux
+        elif tmp == "darwin":
+            return Platform.MacOS
+        else:
+            raise Exception(f"Unknown platform label '{sys.platform}'")
+
+
+class SimpleLog:
+    def __init__(self, filename):
+        userdir = ""
+        platform = Platform.get()
+        if platform == Platform.Windows:
+            userdir = os.environ["USERPROFILE"]
+        elif platform == Platform.Linux:
+            userdir = os.environ["HOME"] if "HOME" in os.environ else ""
+        self.filename = filename.replace(
+            "${userdir}", userdir
+        )  # replace possible ${userdir} placeholder
+
+    def _output_msg(self, level, msg):
+        with open(self.filename, "a") as f:
+            f.write(f"{level:8} {datetime.now()} {msg}\n")
+
+    def info(self, msg):
+        self._output_msg("info", msg)
+
+    def warning(self, msg):
+        self._output_msg("warning", msg)
+
+    def error(self, msg):
+        self._output_msg("error", msg)
+
+    def debug(self, msg):
+        self._output_msg("debug", msg)
+
+
+class ShellCommandReceive:
+    """
+    Helper class for receiving and performing shell commands in a platform independent form
+
+    All supported shell commands have a cmd_ prefix. When adding new functions make sure that the is an
+    equivalent in the ShellCommandSend class. When a command failed a non zero exit code will be returned.
+    Hence, the ShellCommandSend class always uses subprocess.check_output for assessing if the command was
+    successfull. Some command require information to be returned (cmd_exists, cmd_start, cmd_running) which
+    is written to stdout in the following form: __<key>=<value>__
+    """
+
+    def _log(self, msg):
+        if not self.log:
+            return
+        self.log.info(f"[id={self.ranid}] {msg}")
+
+    def __init__(self, debugging=False, use_break_way=True, log=None):
+        self.debugging = debugging
+        self.log = None
+        self.platform = Platform.get()
+        self.use_break_way = use_break_way
```

Comment:
Looks like this should be `'breakaway`, not `break_way` to match what it's referring to

Reply:
You are right! Fix in commit https://github.com/ipython/ipyparallel/commit/0a0efb02be1da21c224f9e805159c13809902e4f